### PR TITLE
expression: implement is_ipv4 and is_ipv6 function pushdown

### LINKED ISSUE #41172: Implement is_ipv4 and is_ipv6 function push down to TiFlash
## Enhancement
upport `is_ipv4()` and `is_ipv6()` function in TiFlash
For details: https://github.com/pingcap/tiflash/pull/6773

### PR DESCRIPTION
**Pull request by AntiTopQuark**
### What problem does this PR solve?

Issue Number: close #41172 
related issue: https://github.com/pingcap/tiflash/pull/6773
Problem Summary:
implement is_ipv4 and is_ipv6 function pushdown
### What is changed and how it works?
1. open pushdown of  is_ipv4 and is_ipv6
2. add some test
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
implement is_ipv4 and is_ipv6 function pushdown to tiflash
```

### CODE REVIEW

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -8431,3 +8431,93 @@ func TestIssue40285(t *testing.T) {
...
+	tk.MustExec("insert into t values('192.168.0.0/10', 'fe80::2dc3:25a5:49a1:6002%24')")
+	tk.MustExec("insert into t values('192.168.99.22.123', '4207:A33A:58D3:F2C3:8EDC:A548:3EC7:0D00:0D00')")
+	tk.MustExec("insert into t values('999.999.999.999', '4207:A33A:58D3:F2C3:8EDC:A548::0D00')")
+	tk.MustExec("insert into t values('3.2.1.', '4207::::8EDC:A548:3EC7:0D00')")
+	tk.MustExec("insert into t values('3..2.1', '4207:::::A548:3EC7:0D00')")
+	tk.MustExec("insert into t values('...', '::::::')")
+	tk.MustExec("insert into t values('4556456', '4556456')")
+	tk.MustExec("insert into t values('ajdjioa', 'ajdjioa')")
+	tk.MustExec("insert into t values('', '')")
+	tk.MustExec("insert into t values(null,null)")
```
Comment (SeaRise) [inline planner/core/integration_test.go:8453]: Seems useless here

File: planner/core/integration_test.go [hunk 2]
```diff
@@ -8431,3 +8431,93 @@ func TestIssue40285(t *testing.T) {
...
```
Comment (SeaRise) [inline planner/core/integration_test.go:8498]: ditto

File: planner/core/integration_test.go [hunk 3]
```diff
@@ -8431,3 +8431,93 @@ func TestIssue40285(t *testing.T) {
 	tk.MustExec("CREATE TABLE t(col1 enum('p5', '9a33x') NOT NULL DEFAULT 'p5',col2 tinyblob DEFAULT NULL) ENGINE = InnoDB DEFAULT CHARSET = latin1 COLLATE = latin1_bin;")
 	tk.MustQuery("(select last_value(col1) over () as r0 from t) union all (select col2 as r0 from t);")
 }
+
+func TestIsIPv4ToTiFlash(t *testing.T) {
```
Comment (SeaRise) [inline planner/core/integration_test.go:8435]: unit test in `expression/expr_to_pb_test.go` need to be added too, like https://github.com/pingcap/tidb/pull/37589/files#diff-6d884fa56c800263f08f60430b9a757c8c536ebfe98de09ec031f89680147765R1181-R1185

Comment (AntiTopQuark) [reply planner/core/integration_test.go:8435]: done,thx~

**Review by LittleFall — APPROVED** [review]

**Review by guo-shaoge — APPROVED** [review]

**Comment by guo-shaoge** [comment]
Please fix case: **TestIsIPv6ToTiFlash** @AntiTopQuark 
```
25665 [2023/02/15 15:56:48.047 +08:00] [WARN] [2pc.go:1738] ["schemaLeaseChecker is not set for this transaction"] [sessionID=10] [startTS=439470734192672770] [checkTS=439470734192672771]
25666     result.go:131:
25667             Error Trace:    /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/7660/execroot/__main__/bazel-out/k8-fastbuild/bin/planner/core/core_test_/core_test.runfiles/__main__/planner/core/result.go:131
25668                                         /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/7660/execroot/__main__/bazel-out/k8-fastbuild/bin/planner/core/core_test_/core_test.runfiles/__main__/planner/core/integration_test.go:8523
25669             Error:          Not equal:
25670                             expected: "[[TableReader_9 root data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [  └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]"
25671                             actual  : "[[TableReader_9 root MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [  └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]"
25672
25673                             Diff:
25674                             --- Expected
25675                             +++ Actual
25676                             @@ -1 +1 @@
25677                             -[[TableReader_9 root data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [  └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]
25678                             +[[TableReader_9 root MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [  └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]
25679             Test:           TestIsIPv6ToTiFlash
25680             Messages:       sql:explain select is_ipv6(v6) from t;, args:[]
```

**Comment by guo-shaoge** [comment]
Looks like case still failed: @AntiTopQuark 
```
[checkTS=439472502008184835]
    result.go:131:
            Error Trace:    /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/7919/execroot/__main__/bazel-out/k8-fastbuild/bin/planner/ core/core_test_/core_test.runfiles/__main__/planner/core/result.go:131
                                        /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/7919/execroot/__main__/bazel-out/k8-fastbuild/ bin/planner/core/core_test_/core_test.runfiles/__main__/planner/core/integration_test.go:8523
            Error:          Not equal:
                            expected: "[[TableReader_9 root   MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [     └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]"
                            actual  : "[[TableReader_9 root MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [       └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]"

                            Diff:
                            --- Expected
                            +++ Actual
                            @@ -1 +1 @@
                            -[[TableReader_9 root   MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [               └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]                                                                    +[[TableReader_9 root MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [  └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]
            Test:           TestIsIPv6ToTiFlash
            Messages:       sql:explain select is_ipv6(v6) from t;, args:[]
```

**Comment by guo-shaoge** [comment]
You can run `go test . --tags=intest -run TestIsIPv6ToTiFlash` to run case in your local evn.

**Comment by AntiTopQuark** [comment]
> You can run `go test . --tags=intest -run TestIsIPv6ToTiFlash` in your local env, and make sure case is ok. @AntiTopQuark

thank u very much, now it is okey.
image

**Comment by AntiTopQuark** [comment]
> Looks like case still failed: @AntiTopQuark
> 
> ```
> [checkTS=439472502008184835]
>     result.go:131:
>             Error Trace:    /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/7919/execroot/__main__/bazel-out/k8-fastbuild/bin/planner/ core/core_test_/core_test.runfiles/__main__/planner/core/result.go:131
>                                         /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/7919/execroot/__main__/bazel-out/k8-fastbuild/ bin/planner/core/core_test_/core_test.runfiles/__main__/planner/core/integration_test.go:8523
>             Error:          Not equal:
>                             expected: "[[TableReader_9 root   MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [     └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]"
>                             actual  : "[[TableReader_9 root MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [       └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]"
> 
>                             Diff:
>                             --- Expected
>                             +++ Actual
>                             @@ -1 +1 @@
>                             -[[TableReader_9 root   MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [               └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]                                                                    +[[TableReader_9 root MppVersion: 1, data:ExchangeSender_8] [└─ExchangeSender_8 mpp[tiflash] ExchangeType: PassThrough] [  └─Projection_4 mpp[tiflash] is_ipv6(test.t.v6)->Column#4] [    └─TableFullScan_7 mpp[tiflash] keep order:false, stats:pseudo]]
>             Test:           TestIsIPv6ToTiFlash
>             Messages:       sql:explain select is_ipv6(v6) from t;, args:[]
> ```

thank u very much~

**Comment by AntiTopQuark** [comment]
it looks like that the test named "check_dev_2" is unstable .
The same problem also occurs with another PR :https://github.com/pingcap/tidb/pull/41453
https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fghpr_check2/detail/ghpr_check2/14262/

**Comment by AntiTopQuark** [comment]
> **idc-jenkins-ci-tidb/check_dev_2 ** Pending — Jenkins job enqueued.                   BaseSHA:3d8b7c21f15b1d45d05169d1c81cd5e42233c894

image

**Pull request merged**